File: src/types/review.ts
/** Shared review domain types — UI + future API contract. */

export type ProductReview = {
  id: string;
  author: string;
  initials: string;
  /** 1–5 */
  rating: number;
  title: string;
  body: string;
  /** Display date, e.g. "Mar 2026" */
  dateLabel: string;
  /** ISO date for stable sorting (API-ready) */
  createdAt: string;
  verified: boolean;
  helpful: number;
  tags?: string[];
  featured?: boolean;
};

export type RatingBreakdown = {
  stars: 1 | 2 | 3 | 4 | 5;
  /** 0–100 share of reviews */
  percent: number;
  count: number;
};

export type ReviewSort = "featured" | "helpful" | "recent";

/** Request shape — same fields a future REST/GraphQL endpoint would accept. */
export type ReviewsQuery = {
  productId: string;
  /** 1-based page index */
  page: number;
  pageSize: number;
  sort: ReviewSort;
  /** Optional topic filter */
  tag?: string | null;
};

export type ReviewsSummary = {
  average: number;
  count: number;
  breakdown: RatingBreakdown[];
  /** Highlight for pull-quote (independent of current page) */
  featured: ProductReview | null;
  /** Distinct tags across the product’s reviews */
  tags: string[];
};

/**
 * Paginated response — designed so the mock layer and a real API
 * return the same shape. UI never pages client-side over “all” reviews.
 */
export type ReviewsPage = {
  items: ProductReview[];
  page: number;
  pageSize: number;
  totalItems: number;
  totalPages: number;
  hasNext: boolean;
  hasPrev: boolean;
  summary: ReviewsSummary;
  /** Echo of the query that produced this page (debugging / sync) */
  query: ReviewsQuery;
};

File: backend/app/api/v1/schemas/review.py
"""Review page schemas — mirrors src/types/review.ts."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

ReviewSort = Literal["featured", "helpful", "recent"]


class CamelModel(BaseModel):
    model_config = ConfigDict(
        populate_by_name=True,
        ser_json_by_alias=True,
    )


class ProductReviewOut(CamelModel):
    id: str
    author: str
    initials: str
    rating: int
    title: str
    body: str
    date_label: str = Field(serialization_alias="dateLabel")
    created_at: str = Field(serialization_alias="createdAt")
    verified: bool
    helpful: int
    tags: list[str] | None = None
    featured: bool | None = None


class RatingBreakdownOut(CamelModel):
    stars: Literal[1, 2, 3, 4, 5]
    percent: int
    count: int


class ReviewsQueryOut(CamelModel):
    product_id: str = Field(serialization_alias="productId")
    page: int
    page_size: int = Field(serialization_alias="pageSize")
    sort: ReviewSort
    tag: str | None = None


class ReviewsSummaryOut(CamelModel):
    average: float
    count: int
    breakdown: list[RatingBreakdownOut]
    featured: ProductReviewOut | None
    tags: list[str]


class ReviewsPageOut(CamelModel):
    items: list[ProductReviewOut]
    page: int
    page_size: int = Field(serialization_alias="pageSize")
    total_items: int = Field(serialization_alias="totalItems")
    total_pages: int = Field(serialization_alias="totalPages")
    has_next: bool = Field(serialization_alias="hasNext")
    has_prev: bool = Field(serialization_alias="hasPrev")
    summary: ReviewsSummaryOut
    query: ReviewsQueryOut
